File: tests/data_structures/linked_list/linked_list.py
"""
Linked List benchmark implementation in Python.
Tests basic linked list operations: insert, search, delete.
"""

import time


class ListNode:
    """Node for linked list."""
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class LinkedList:
    """Simple linked list implementation."""
    
    def __init__(self):
        self.head = None
        self.size = 0
    
    def insert(self, val):
        """Insert value at the beginning."""
        new_node = ListNode(val)
        new_node.next = self.head
        self.head = new_node
        self.size += 1
    
    def search(self, val):
        """Search for a value in the list."""
        current = self.head
        position = 0
        while current:
            if current.val == val:
                return position
            current = current.next
            position += 1
        return -1
    
    def delete(self, val):
        """Delete first occurrence of value."""
        if not self.head:
            return False
        
        if self.head.val == val:
            self.head = self.head.next
            self.size -= 1
            return True
        
        current = self.head
        while current.next:
            if current.next.val == val:
                current.next = current.next.next
                self.size -= 1
                return True
            current = current.next
        return False
    
    def get_size(self):
        """Get the size of the list."""
        return self.size


def main():
    """Run linked list benchmark."""
    print("Starting linked list benchmark...")
    start_time = time.perf_counter()
    
    # Create linked list and perform operations
    linked_list = LinkedList()
    operations_count = 10000
    
    # Insert operations
    for i in range(operations_count):
        linked_list.insert(i)
    
    # Search operations
    found_count = 0
    for i in range(0, operations_count, 100):  # Search every 100th element
        if linked_list.search(i) != -1:
            found_count += 1
    
    # Delete operations
    deleted_count = 0
    for i in range(0, operations_count, 200):  # Delete every 200th element
        if linked_list.delete(i):
            deleted_count += 1
    
    end_time = time.perf_counter()
    execution_time = end_time - start_time
    
    print(f"Operations completed: {operations_count} inserts, {found_count} searches, {deleted_count} deletes")
    print(f"Final list size: {linked_list.get_size()}")
    print(f"Execution time: {execution_time:.6f} seconds")


if __name__ == "__main__":
    main()

File: tests/data_structures/linked_list/linked_list.ts
/**
 * Linked List benchmark implementation in TypeScript.
 * Tests basic linked list operations: insert, search, delete.
 */

class ListNode {
    val: number;
    next: ListNode | null;
    
    constructor(val: number) {
        this.val = val;
        this.next = null;
    }
}

class LinkedList {
    private head: ListNode | null;
    private size: number;
    
    constructor() {
        this.head = null;
        this.size = 0;
    }
    
    insert(val: number): void {
        const newNode = new ListNode(val);
        newNode.next = this.head;
        this.head = newNode;
        this.size++;
    }
    
    search(val: number): number {
        let current = this.head;
        let position = 0;
        
        while (current !== null) {
            if (current.val === val) {
                return position;
            }
            current = current.next;
            position++;
        }
        return -1;
    }
    
    delete(val: number): boolean {
        if (this.head === null) {
            return false;
        }
        
        if (this.head.val === val) {
            this.head = this.head.next;
            this.size--;
            return true;
        }
        
        let current = this.head;
        while (current.next !== null) {
            if (current.next.val === val) {
                current.next = current.next.next;
                this.size--;
                return true;
            }
            current = current.next;
        }
        return false;
    }
    
    getSize(): number {
        return this.size;
    }
}

function main(): void {
    console.log("Starting linked list benchmark...");
    const startTime = process.hrtime.bigint();
    
    const linkedList = new LinkedList();
    const operationsCount = 10000;
    
    // Insert operations
    for (let i = 0; i < operationsCount; i++) {
        linkedList.insert(i);
    }
    
    // Search operations
    let foundCount = 0;
    for (let i = 0; i < operationsCount; i += 100) {
        if (linkedList.search(i) !== -1) {
            foundCount++;
        }
    }
    
    // Delete operations
    let deletedCount = 0;
    for (let i = 0; i < operationsCount; i += 200) {
        if (linkedList.delete(i)) {
            deletedCount++;
        }
    }
    
    const endTime = process.hrtime.bigint();
    const executionTime = Number(endTime - startTime) / 1e9; // Convert to seconds
    
    console.log(`Operations completed: ${operationsCount} inserts, ${foundCount} searches, ${deletedCount} deletes`);
    console.log(`Final list size: ${linkedList.getSize()}`);
    console.log(`Execution time: ${executionTime.toFixed(6)} seconds`);
}

main();